### ISSUE TITLE
Vertical off-center writes convenience

### ISSUE THREAD

**Issue by rheacangeo** [GridTools/gt4py#131]
Some column based atmospheric science algorithms are intuitive to scientists to write in such a way that adjacent vertical layers are modified in conjunction. For example, if you take mass from the layer you are at and add it to the one below you, it is intuitive to write:
```with computation(BACKWARD), interval(1, None):
        if mass[0, 0, 0] > 0:
             delta = x * y
             mass[0, 0, 0]  = mass[0, 0, 0]  - delta
             mass[0, 0, -1] = mass[0, 0, -1] + delta ```
Without off-center writes, we'd need to write it something like
with computation(BACKWARD), interval(1, None):
        delta = x * y
        if mass[0, 0, 1] > 0:
               mass = mass + delta[0, 0, 1]
        if mass > 0:
               mass  = mass  - delta
This is not terrible, and makes sense with this small example, but it quickly can get confusing with more complex patterns.   
Here is a sample we ported:    
# Python loop version
for j in range(js, je ):
    for k in range(1, nz - 1):
        for i in range(is_, ie):
            if qv[i, j, k] < 0 and qv[i, j, k - 1] > 0.0:
                dq = min(
                    -qv[i, j, k] * dp[i, j, k], qv[i, j, k - 1] * dp[i, j, k - 1]
                )
                qv[i, j, k - 1] -= dq / dp[i, j, k - 1]
                qv[i, j, k] += dq / dp[i, j, k]
            if qv[i, j, k] < 0.0:
                qv[i, j, k + 1] += qv[i, j, k] * dp[i, j, k] / dp[i, j, k + 1]
                qv[i, j, k] = 0.0

# Stencil version avoiding offcenter writes-introduce fields upper_fix and lower_fix:
@utils.stencil()
def fix_water_vapor_down(qv: sd, dp: sd, upper_fix: sd, lower_fix: sd, dp_bot: sd):
    with computation(PARALLEL):
        with interval(1, 2):
            if qv[0, 0, -1] < 0:
                qv = qv + qv[0, 0, -1] * dp[0, 0, -1] / dp
        with interval(0, 1):
            qv = qv if qv >= 0 else 0
    with computation(FORWARD), interval(1, -1):
        dq = qv[0, 0, -1] * dp[0, 0, -1]
        if lower_fix[0, 0, -1] != 0:
            qv = qv + lower_fix[0, 0, -1] / dp
        if (qv < 0) and (qv[0, 0, -1] > 0):
            dq = dq if dq < -qv * dp else -qv * dp
            upper_fix = dq
            qv = qv + dq / dp
        if qv < 0:
            lower_fix = qv * dp
            qv = 0
    with computation(PARALLEL), interval(0, -2):
        if upper_fix[0, 0, 1] != 0:
            qv = qv - upper_fix[0, 0, 1] / dp
    with computation(PARALLEL), interval(-1, None):
        if lower_fix[0, 0, -1] > 0:
            qv = qv + lower_fix / dp
        upper_fix = qv

In the stencil version, it's hard to tell what's going on. The examples can get more complex.

**Comment by jdahm** [comment]
We discussed this in a team meeting and decided that while this would be quite useful for certain vertical solve patterns, there are other open issues and feature requests that will have much higher impact for the work required (for example, directional offsets).

So, I suggest we do not include this in the effort for the v1.0 release, but also not close the issue.

**Comment by gronerl** [comment]
Should the last update be to `temperature[m]` ? Like this it looks like the second last update will be overwritten. Or is another index wrong?

**Comment by oelbert** [comment]
No, that's correct it should be `temperature[m]`, fixed now

**Comment by gronerl** [comment]
If in the end, DaCe orchestration is the goal, another solution would be to write a NumPy code and directly use `@dace.program` which will integrate nicely into the DaCe orchestrated context.

**Comment by havogt** [comment]
@oelbert
- Would it be a lot of work to make a standalone testcase for your example in Numpy? I guess the hard-work is to initialize the arrays with data that triggers the relevant codepaths.
- Would the code be more or less appealing if we split "removing ice/snow/graupel from one k-level and adding it as rain to the appropriate lower levels" and "updating the heat capacities and temperatures along the way" into separate computations? Trying to figure out how much of the shape of the code is due to optimizations.
- What is `r1`?

**Comment by oelbert** [comment]
Currently this is just a Python function to be dace-optimized, though I do have a standalone test for it. I've been serializing data from Fortran runs and  checking the coverage in Python. It works, but it does feel awkward to have the model be in GT4Py except for this one piece of code that's not. I think Linus' workaround could solve putting this into stencil, though.

As far as splitting the heat capacity and temperature updates from the sedimentation melting, that would be ideal except for the `sink = min(qice[k] * delp[k] / delp[m], dtime * (temperature[m] - constants.TICE) / icpk[m])` line, where the amount of ice taken from level k depends in part on the temperature at level m, which updates through the loop.

`r1` is an i-j field that tracks the rain that's precipitated to the ground through the sedimentation scheme.

**Comment by romanc** [comment]
@FlorianDeconinck is this issue resolved with "k offset write" merged in #1452? I know, we still have [issues with k offset writes on GPUs](https://github.com/GridTools/gt4py/pull/1800). The CPU version is solid afaik. Maybe we could close this issue and open a dedicated one for the GPU issues?

**Comment by FlorianDeconinck** [comment]
Yes - we need a good repro though and the deactivated case is not a good repro since we can reproduce it locally.
